src/odemis/driver/technolution.py

```diff
@@ -0,0 +1,1449 @@
+        self._mppc = mppc
+
+    def start_generate(self):
+        """
+        Start the dataflow using the provided function. The appropriate settings are retrieved via the VA's of the
+        each component
+        """
+        self._mppc.startAcquisition()
+
+    def next(self, field_num):
+        """
+        Acquire the next field image using the provided function.
+        :param field_num (tuple): tuple with x,y coordinates in integers of the field number.
+        """
+        self._mppc.getNextField(field_num)
+
+    def stop_generate(self):
+        """
+        Stop the dataflow using the provided function.
+        """
+        self._mppc.stopAcquisition()
+
+    def get(self, *args, **kwargs):
+        """
+        Acquire a single field, can only be called if no other acquisition is active.
+        :return: (DataArray)
+        """
+        if self._count_listeners() < 1:
+            # Acquire and return received image
+            image = self._mppc.acquireSingleField(*args, **kwargs)
+            return image
+
+        else:
+            logging.error("There is already an acquisition on going with %s listeners subscribed, first cancel/stop "
+                          "current running acquisition to acquire a single field-image" % self._count_listeners())
+            raise Exception("There is already an acquisition on going with %s listeners subscribed, first cancel/stop "
+                            "current running acquisition to acquire a single field-image" % self._count_listeners())
+
+
+class AsmApiException(Exception):
+    """
+    Exception for raising errors while calling the ASM API.
+    """
+
+    def __init__(self, url, response, expected_status, ASM):
+        """
+        Initializes exception object defining the error message to be displayed to the user as a response.
+        And performs basic checks on ASM items to see of those are not the cause of the error. (e.g. monitor/item
+        sam_connection_operational, ext_store_connection_operational, offload_queue_fill_level, install_in_progress,
+        last_install_success)
+
+        :param url: URL of the call tried which was tried to make
+        :param response: full/raw response from the ASM API
+        :param expected_status: the expected status code
+        :param ASM (AcquisitionServer object) : AcquisitionServer object used for changed the state of the system and
+        checking system parameters.
+        """
+        self.url = url
+        self.status_code = response.status_code
+        self.reason = response.reason
+        self.expected_status = expected_status
+
+        # TODO Currently the system/"monitor" checks in "system_checks(ASM) are not supported in the simulator,
+        #  they are only defined in the  API. Therefore this method is commented to prevent creating errors.
+        # self.system_checks(ASM)  # Perform general checks on the ASM system to find the cause of the error.
```

Comment:
> BTW, the exception also has to be "pickable" so that it can be sent to the Odemis client. I'm not sure a URL object is pickable, and anyway you don't need it after init. So I'd recommend to not hold url as an attribute. Just pass it to the functions as a parameter.

I just saw that the docstring doesn't include the types of the parameters (which I will include). The URL here is only the string URL. Which probably is pickable. I do think it is important to use that string since it is used to display the error message and which call caused the error.

Reply:
Ok, in any case, it should be possible to compute the error message in `__init__()`, you don't need to keep the arguments as attributes. Only keep in attributes the arguments which are useful for processing the exception (eg, an error code).